plugin/mmdb/setup.go

```diff
@@ -0,0 +1,94 @@
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/oschwald/geoip2-golang"
+)
+
+const pluginName = "mmdb"
+
+func init() { plugin.Register(pluginName, setup) }
+
+func setup(c *caddy.Controller) error {
+	mmdb, err := mmdbParse(c)
+	if err != nil {
+		return plugin.Error(pluginName, err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		mmdb.Next = next
+		return mmdb
+	})
+
+	return nil
+}
+
+// probingIP is an arbitrary internet IP used to discover database types through database lookups.
+var probingIP = net.ParseIP("81.2.69.142")
+
+func mmdbParse(c *caddy.Controller) (*Mmdb, error) {
+	dbPaths := make(map[string]bool)
+	for c.Next() {
+		if !c.NextArg() {
+			return nil, c.ArgErr()
+		}
+		path := c.Val()
+		if c.NextArg() { // There shouldn't be any more arguments.
+			return nil, c.ArgErr()
+		}
+		if dbPaths[path] {
+			return nil, c.Errf("repeating mmdb path %q is not allowed", path)
+		}
+		dbPaths[path] = true
+		c.NextBlock()
+	}
+
+	// TODO(snebel29): Warn or Fail when configuring overlapping db/fields?
+	// TODO(snebel29): Fail if we open more than once the same dbType?
+	dbs := make([]mmdb, 0, len(dbPaths))
+
+	for path := range dbPaths {
+		reader, err := geoip2.Open(path)
+		if err != nil {
+			return nil, plugin.Error(pluginName, c.Errf("failed to open database file: %v", err))
+		}
+		db := mmdb{Reader: reader}
+		schemas := []struct {
+			provides int
+			name     string
+			validate func() error
+		}{
+			{name: "enterprise", provides: enterprise, validate: func() error { _, err := reader.Enterprise(probingIP); return err }},
+			{name: "city", provides: city, validate: func() error { _, err := reader.City(probingIP); return err }},
+			{name: "country", provides: country, validate: func() error { _, err := reader.Country(probingIP); return err }},
+			{name: "asn", provides: asn, validate: func() error { _, err := reader.ASN(probingIP); return err }},
+			{name: "connectionType", provides: connectionType, validate: func() error { _, err := reader.ConnectionType(probingIP); return err }},
+			{name: "domain", provides: domain, validate: func() error { _, err := reader.Domain(probingIP); return err }},
+			{name: "isp", provides: isp, validate: func() error { _, err := reader.ISP(probingIP); return err }},
+			{name: "anonymousIP", provides: anonymousIP, validate: func() error { _, err := reader.AnonymousIP(probingIP); return err }},
+		}
+		// Probe all databases to figure out the database information that it can provide.
+		for _, schema := range schemas {
+			if err := schema.validate(); err != nil {
+				// If we get an InvalidMethodError then we know this database does not provide enterprise schema.
+				if _, ok := err.(geoip2.InvalidMethodError); !ok {
+					return nil, plugin.Error(pluginName,
+						c.Errf("unexpected failure looking up database %q schema %q: %v", filepath.Base(path), schema.name, err))
+				}
+			} else {
```

Comment:
danling else, can just to continue in the above if and remove the else entirely

Reply:
Will do.